Feature/fe fix load error redux

maismin opened the pull request:
- [x] Create loading reducer to handle all API request states

- [x] Create error reducer to handle API error messages

- [x] Fix sign-up and sign-in components to use the loading and error reducers

maismin committed:
Add loader reducer to store all API request states

- This prevents polluting the reducers with loading flags everywhere.
Now the reducers can focus on storing data.

- The createLoadingSelector is used to access the loading states in the
component.

maismin committed:
Fix names of action types to be more uniform

- This makes the action types easier to read.

- Refactor auth reducer and auth actions to use the new action types

maismin committed:
Add CircularProgress component to display loading

- Fix ThemeButton component to show progress instead of text when an API
requeste is made.

- Add CircularProgress component to App, when the user first visits the
page and is authenticated.

maismin committed:
Fix ThemeButton.js

- Remove loading attribute from Button

maismin committed:
Fix auth reducer

- Remove action types with REQUEST suffixes as they are used for setting
the loading state, and the auth reducer does not handle the loading
states.

maismin committed:
Fix authenticateUser() to return correct payload

maismin committed:
Add error reducer to handle API error messages

in one place, reducing code for handling error states in multiple
reducers.

maismin committed:
Add error handling for login/register components

- The login and register components have separate error states, and
previous errors are cleared before rendering the components.

Esodey commented on client/src/api/selectors.js:
```diff
@@ -0,0 +1,12 @@
+export const createLoadingSelector = actions => state => {
+  // returns true only when all actions is not loading
```
rephrasing comment would be helpful

maismin committed:
Document error and loading reducers

- Add comments for the higher order functions in the selectors.js
file.

- Fix comment formatting for the reducers.

- Add comments in the SignIn and SignUp forms to facilitate better
understanding of the higher order functions' usage.

maismin committed:
Fix PrivateRoute

- Remove loading state from PrivateRoute

maismin committed:
Refactor error and loading selectors

- Move the error and loading selectors into the error and loading
reducers respectively as this is considered a Redux best practice.

- Fix components to import the error and loading selectors correctly.

- Fix the circular progress component in the ThemeButton component to be
in the center.
